### PR TITLE
feat:[CDS-68602]: Custom Sequence changes with Junits for service dashoard

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
These changes corresponds to custom sequencing in custom dashboard with Junits for service dashboard.

Tests Done:
1. use of custom sequence in card view
2. use of default sequence in card view
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47345)

### CODE REVIEW

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 1]
```diff
@@ -3328,7 +3383,7 @@ private List<EnvironmentGroupInstanceDetails.EnvironmentGroupInstanceDetail> get
 
       HashMap<String, CustomSequenceDTO.EnvAndEnvGroupCard> envGrpMapForCustomSequence = new HashMap<>();
 
-      envAndEnvGroupCards.forEach(envGroupDetail
+      envAndEnvGroupCardsCustom.forEach(envGroupDetail
           -> envGrpMapForCustomSequence.put(
               envGroupDetail.getName() + envGroupDetail.getIdentifier() + envGroupDetail.isEnvGroup(), envGroupDetail));
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:3389]: we don't need to add name to key

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 2]
```diff
@@ -3342,14 +3397,12 @@ private List<EnvironmentGroupInstanceDetails.EnvironmentGroupInstanceDetail> get
           appendListForEnvGrpNotPresentInSequence.add(envGrpMapForListFromDB.get(key));
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:3397]: we can use entry set to traverse

Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:3397]: can use contains key

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 3]
```diff
@@ -1616,73 +1616,123 @@ public PipelineExecutionCountInfo getPipelineExecutionCountInfo(String accountId
...
     Optional<ServiceSequence> serviceSequenceOptional =
         serviceSequenceService.get(accountIdentifier, orgIdentifier, projectIdentifier, serviceId);
     ServiceSequence serviceSequence;
-    if (serviceSequenceOptional.isPresent()) {
-      serviceSequence = serviceSequenceOptional.get();
-    } else {
-      throw new InvalidRequestException("No default sequence of Env and Env Group exist for this service");
+    if (!serviceSequenceOptional.isPresent()) {
+      // message
+      return getSequenceDTO(
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1626]: we can call this method and save data in a variable and return that directly

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 4]
```diff
@@ -1616,73 +1616,123 @@ public PipelineExecutionCountInfo getPipelineExecutionCountInfo(String accountId
...
+    List<String> defaultKeys = new ArrayList<>();
+    List<String> customKeys = new ArrayList<>();
+    List<CustomSequenceDTO.EnvAndEnvGroupCard> newEnvAndEnvGroupCardList = new ArrayList<>();
+    List<CustomSequenceDTO.EnvAndEnvGroupCard> appendEnvAndEnvGroupCardList = new ArrayList<>();
 
-    for (String envKey : envGrpCardsMap.keySet()) {
-      CustomSequenceDTO.EnvAndEnvGroupCard customCard = envGrpCardsMap.get(envKey);
+    defaultSequence.getEnvAndEnvGroupCardList().forEach(
+        card -> defaultKeys.add(card.getIdentifier() + card.getName() + card.isEnvGroup()));
+    customSequence.getEnvAndEnvGroupCardList().forEach(
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1651]: avoid name in key

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 5]
```diff
@@ -1616,73 +1616,123 @@ public PipelineExecutionCountInfo getPipelineExecutionCountInfo(String accountId
...
+        card -> customKeys.add(card.getIdentifier() + card.getName() + card.isEnvGroup()));
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1653]: here also

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 6]
```diff
@@ -1616,73 +1616,123 @@ public PipelineExecutionCountInfo getPipelineExecutionCountInfo(String accountId
...
 
-      boolean isNew = customCard.isNew();
-      customCard.setNew(false);
+    customSequence.getEnvAndEnvGroupCardList().forEach(card -> {
+      if (defaultKeys.contains(card.getIdentifier() + card.getName() + card.isEnvGroup())) {
+        newEnvAndEnvGroupCardList.add(card);
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1656]: defaultKeys/customKeys please create set in place of list

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 7]
```diff
@@ -1616,73 +1616,123 @@ public PipelineExecutionCountInfo getPipelineExecutionCountInfo(String accountId
...
+      }
+    });
 
-      if (!defaultSequence.contains(customCard)) {
-        customSequence.remove(customCard);
+    defaultSequence.getEnvAndEnvGroupCardList().forEach(card -> {
+      if (!customKeys.contains(card.getIdentifier() + card.getName() + card.isEnvGroup())) {
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1661]: we can traverse defaultSequece and have logic if it custom contains it then push else set new to true and push.

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 8]
```diff
@@ -1616,73 +1616,123 @@ public PipelineExecutionCountInfo getPipelineExecutionCountInfo(String accountId
...
+        card.setNew(true);
+        appendEnvAndEnvGroupCardList.add(card);
       }
-      customCard.setNew(isNew);
+    });
+
+    newEnvAndEnvGroupCardList.addAll(0, appendEnvAndEnvGroupCardList);
+    return CustomSequenceDTO.builder().envAndEnvGroupCardList(newEnvAndEnvGroupCardList).build();
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1668]: append to appendEnvAndEnvGroupCardList directly and return

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 9]
```diff
@@ -1616,73 +1616,123 @@ public PipelineExecutionCountInfo getPipelineExecutionCountInfo(String accountId
...
+      return sequenceDTO;
+    }
+    CustomSequenceDTO customSequenceDTO = serviceSequenceOptional.get().getCustomSequence();
+
+    if (isNull(customSequenceDTO)) {
+      return sequenceDTO;
     }
+    Map<String, Boolean> cardToIsNew = new HashMap<>();
+    customSequenceDTO.getEnvAndEnvGroupCardList().forEach(
+        card -> cardToIsNew.put(card.getIdentifier() + card.getName() + card.isEnvGroup(), card.isNew()));
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1689]: remove name for key

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger codeformat

**Comment by sourabh-awashti** [comment]
trigger codebasehash

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger codeformat

**Comment by sourabh-awashti** [comment]
trigger codeformat

**Comment by sourabh-awashti** [comment]
trigger codebasehashcheck

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Comment by sourabh-awashti** [comment]
trigger pmd

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Comment by sourabh-awashti** [comment]
trigger codeformat

**Review by Abhishek-CDC — APPROVED** [review]

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger codeformat, trigger

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by Abhishek-CDC — APPROVED** [review]

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/overview/resource/CDDashboardOverviewResource.java` line 667 at r7](https://reviewable.io/reviews/harness/harness-core/47345#-NVZ8a4RDWUlOPYP7Dpe:-NVZ8a4RDWUlOPYP7Dpf:b-q7kxpf) ([raw file](https://github.com/harness/harness-core/blob/a56507c08143247cf2a983fc4e41eebd94f1d2fe/120-ng-manager/src/main/java/io/harness/ng/overview/resource/CDDashboardOverviewResource.java#L667)):*
> ```Java
>   @GET
>   @Path("/useCustomSequence")
>   @ApiOperation(value = "get the status of current sequence of env cards ", nickname = "getCustomSequenceStatus")
> ```

we can mark these 3 APIs are hidden

**Comment by yogesh-chauhan** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/entity/ServiceSequence.java` line 62 at r7](https://reviewable.io/reviews/harness/harness-core/47345#-NVZ9LCG8rLDk160UAyx:-NVZ9LCG8rLDk160UAyy:b-xh0juw) ([raw file](https://github.com/harness/harness-core/blob/a56507c08143247cf2a983fc4e41eebd94f1d2fe/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/entity/ServiceSequence.java#L62)):*
> ```Java
>   @Trimmed private String projectIdentifier;
>   @Trimmed private String serviceIdentifier;
> ```

// we should use primitive

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 2 of 6 files at r1, 3 of 9 files at r2, 3 of 4 files at r4, 1 of 3 files at r5, 2 of 3 files at r6, 2 of 3 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47345)* status: 14 file(s) to review. 0 of 2 approvals obtained.

**Review by Abhishek-CDC — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by sourabh-awashti** [comment]
trigger review

**Pull request merged**